sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/s3/S3FileSystem.java

```diff
@@ -502,52 +508,81 @@ protected void copy(
 
   @VisibleForTesting
   void copy(S3ResourceId sourcePath, S3ResourceId destinationPath) throws IOException {
-    String uploadId;
-    long objectSize;
     try {
       ObjectMetadata objectMetadata =
           amazonS3.getObjectMetadata(sourcePath.getBucket(), sourcePath.getKey());
-      objectSize = objectMetadata.getContentLength();
-
-      InitiateMultipartUploadRequest initiateUploadRequest =
-          new InitiateMultipartUploadRequest(destinationPath.getBucket(), destinationPath.getKey())
-              .withStorageClass(storageClass)
-              .withObjectMetadata(objectMetadata);
-
-      InitiateMultipartUploadResult initiateUploadResult =
-          amazonS3.initiateMultipartUpload(initiateUploadRequest);
-      uploadId = initiateUploadResult.getUploadId();
-
+      if (objectMetadata.getContentLength() < MAX_COPY_OBJECT_SIZE_BYTES) {
+        atomicCopy(sourcePath, destinationPath);
+      } else {
+        multipartCopy(sourcePath, destinationPath, objectMetadata);
+      }
     } catch (AmazonClientException e) {
       throw new IOException(e);
     }
+  }
 
-    List<PartETag> eTags = new ArrayList<>();
+  @VisibleForTesting
+  void atomicCopy(S3ResourceId sourcePath, S3ResourceId destinationPath)
+      throws AmazonClientException {
+    CopyObjectRequest copyObjectRequest =
+        new CopyObjectRequest(
+            sourcePath.getBucket(),
+            sourcePath.getKey(),
+            destinationPath.getBucket(),
+            destinationPath.getKey());
+    copyObjectRequest.setStorageClass(storageClass);
+
+    amazonS3.copyObject(copyObjectRequest);
+  }
 
-    long bytePosition = 0;
+  @VisibleForTesting
+  void multipartCopy(
+      S3ResourceId sourcePath, S3ResourceId destinationPath, ObjectMetadata objectMetadata)
+      throws AmazonClientException {
+    InitiateMultipartUploadRequest initiateUploadRequest =
+        new InitiateMultipartUploadRequest(destinationPath.getBucket(), destinationPath.getKey())
+            .withStorageClass(storageClass)
+            .withObjectMetadata(objectMetadata);
+
+    InitiateMultipartUploadResult initiateUploadResult =
+        amazonS3.initiateMultipartUpload(initiateUploadRequest);
+    final String uploadId = initiateUploadResult.getUploadId();
+
+    List<PartETag> eTags = new ArrayList<>();
 
-    // Amazon parts are 1-indexed, not zero-indexed.
-    for (int partNumber = 1; bytePosition < objectSize; partNumber++) {
+    final long objectSize = objectMetadata.getContentLength();
+    if (objectSize == 0) {
```

Comment:
Yes if you follow the flow of the copy method this won't happen. However the problem was detected because the multipartCopy did not invoke correctly the Amazon APIs when copying an empty object. The extra validation makes multipart copy robust to deal also files of size 0 so it is worth to have in case someone relies in the future on this method.

Reply:
Ok, so add a comment that says something like "it is an extra validation in case a caller calls directly S3FileSystem.multipartCopy without using S3FileSystem.copy in the future"